File: [0289][Medium][Game_of_Life]/Game_of_Life.cpp
// leetcode time     cost : 4 ms
// leetcode memory   cost : 6.8 MB 
#include <stdlib.h>
#include<iostream>
#include <queue>
#include <algorithm>
using namespace std;
class Solution {
public:
    void gameOfLife(vector<vector<int>>& board) {
        int dx[] = {-1,  0,  1, -1, 1, -1, 0, 1};
        int dy[] = {-1, -1, -1,  0, 0,  1, 1, 1};

        for(int i = 0; i < board.size(); i++) {
            for(int j = 0 ; j < board[0].size(); j++) {
                int sum = 0;
                for(int k = 0; k < 8; k++) {
                    int nx = i + dx[k];
                    int ny = j + dy[k];
                    if(nx >= 0 && nx < board.size() && ny >= 0 && ny < board[0].size()) {
                        sum += (board[nx][ny]&1); // 只累加最低位
                    }
                }
                if(board[i][j] == 1) {
                    if(sum == 2 || sum == 3) {
                        board[i][j] |= 2;  // 使用第二个bit标记是否存活
                    }
                } else {
                    if(sum == 3) {
                        board[i][j] |= 2; // 使用第二个bit标记是否存活
                    }
                }
            }
        }
        for(int i = 0; i < board.size(); i++) {
            for(int j = 0; j < board[i].size(); j++) {
                board[i][j] >>= 1; //右移一位，用第二bit覆盖第一个bit。
            }
        }
    }
};

File: [0289][Medium][Game_of_Life]/Game_of_Life.py
#-*- coding: utf-8 -*-  
# leetcode time     cost : 48 ms
# leetcode memory   cost : 13.7 MB
# Time  Complexity: O(N*M)
# Space Complexity: O(1)
# solution 1, loop
class Solution:
    def gameOfLife(self, board: List[List[int]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        n = len(board)  # row
        m = len(board[0]) # col
        # for current 0, next action die is -2, revive is -3,
        # for current 0, next action die is 2, keep alive is 3,
        # define 8 directions
        directions = [[-1,0],[-1,-1],[0,-1],[1,-1],[1,0],[1,1],[0,1],[-1,1]]
        def isValid(i,j):
            return 0<=i<n and 0<=j<m
        for i in range(n):
            for j in range(m):
                countAlive = 0
                for direct in directions:
                    new_i = i+ direct[0]
                    new_j = j+ direct[1]
                    # count around cell alive
                    if isValid(new_i,new_j) and board[new_i][new_j]>0:
                        countAlive += 1
                # set next action
                if board[i][j] and (countAlive < 2 or countAlive >3):
                    board[i][j] = 2 
                if board[i][j] and (countAlive is 2 or countAlive is 3):
                    board[i][j] = 3
                if board[i][j]==0 and countAlive is 3:
                    board[i][j] = -3
        #print(board)
        for i in range(n):
            for j in range(m):
                board[i][j] = 1 if (board[i][j]==1 or board[i][j]==3 or board[i][j]==-3) else 0